src/Messenger/DI/MessengerExtension.php

```diff
@@ -331,68 +332,127 @@ private function getHandlersForBus(string $busName): array
      */
-    private function getHandlerDefinitions(string $serviceName): iterable
+    private function guessHandledClasses(ReflectionClass $handlerReflection, string $serviceName, string $methodName): iterable
     {
-        $serviceDefinition = $this->getContainerBuilder()->getDefinition($serviceName);
-        $handlerClassName  = $serviceDefinition->getType();
-        $alias             = $serviceDefinition->getTag(self::TAG_HANDLER)['alias'] ?? null;
-        assert(class_exists($handlerClassName));
-        $handlerDefinitions = [];
-
-        if (is_a($handlerClassName, MessageSubscriberInterface::class, true)) {
-            foreach (call_user_func([$handlerClassName, 'getHandledMessages']) as $message => $definition) {
-                $messageName                      = is_string($definition) ? $definition : (string) $message;
-                $handlerDefinitions[$messageName] = new HandlerDefinition(
-                    $serviceName,
-                    $definition['method'] ?? '__invoke',
-                    $alias
-                );
-            }
+        $handlerClassName = $handlerReflection->getName();
 
-            return $handlerDefinitions;
-        }
+        if ($handlerReflection->implementsInterface(MessageSubscriberInterface::class)) {
+            $getHandledMessages = [$handlerClassName, 'getHandledMessages'];
 
-        $handlerReflection = new ReflectionClass($handlerClassName);
+            if (is_callable($getHandledMessages)) {
+                return $getHandledMessages();
+            }
+        }
 
         try {
-            $method = $handlerReflection->getMethod('__invoke');
+            $method = $handlerReflection->getMethod($methodName);
         } catch (ReflectionException $e) {
-            throw InvalidHandlerService::missingInvokeMethod($serviceName, $handlerReflection->getName());
+            throw InvalidHandlerService::missingHandlerMethod($serviceName, $handlerClassName, $methodName);
         }
 
         if ($method->getNumberOfRequiredParameters() !== 1) {
-            throw InvalidHandlerService::wrongAmountOfArguments($serviceName, $handlerReflection->getName());
+            throw InvalidHandlerService::wrongAmountOfArguments($serviceName, $handlerClassName, $methodName);
         }
 
         $parameter     = $method->getParameters()[0];
         $parameterName = $parameter->getName();
         $type          = $parameter->getType();
-        assert($type instanceof ReflectionNamedType || $type === null);
+        assert($type instanceof ReflectionNamedType || $type instanceof ReflectionUnionType || $type === null);
 
         if ($type === null) {
-            throw InvalidHandlerService::missingArgumentType($serviceName, $handlerClassName, $parameterName);
+            throw InvalidHandlerService::missingArgumentType($serviceName, $handlerClassName, $methodName, $parameterName);
+        }
+
+        if ($type instanceof ReflectionUnionType) {
```

Comment:
We should add test for union types (see code coverage https://coveralls.io/builds/54132613/source?filename=src%2FMessenger%2FDI%2FMessengerExtension.php#L434)